gui: Add better support for high DPI displays

Yunotchi opened the pull request:
# About
This PR is carrying on work from #1220.
- Vita3K's window size will scale depending on the user's OS scaling setting -- currently only implemented for Windows
- Fonts will render at a higher resolution to match the larger window size
- Back button's appearance is more consistent across system apps
- Images (e.g. icons and background images) scale nicer
  - Switched to linear filtering from nearest neighbour. This could have knock-on effects that I am not aware of; please check my change to `vita3k\gui\src\imgui_impl_sdl_gl3.cpp`
- App selector's icons now scale with window size
- Content manager: black bar at bottom now fills the whole window's width

# Screenshots
Open in separate tabs to compare before and after.

| Before | After |
| --- | --- |
| <img  src="https://user-images.githubusercontent.com/41648364/114246019-807a4c80-9989-11eb-8f0c-84479982ad0d.png"> | <img src="https://user-images.githubusercontent.com/41648364/114246023-82441000-9989-11eb-9656-4837c7051855.png"> |
| <img src="https://user-images.githubusercontent.com/41648364/114246158-cdf6b980-9989-11eb-922d-a2e947198649.png"> | <img src="https://user-images.githubusercontent.com/41648364/114246155-cd5e2300-9989-11eb-936d-28e0d12c5bf8.png"> |
| <img src="https://user-images.githubusercontent.com/41648364/114246604-ef0bda00-998a-11eb-89be-80db317de9cf.png"> | <img src="https://user-images.githubusercontent.com/41648364/114319139-1340f580-9b08-11eb-8c5b-332b4ba9c00e.png"> |
| <img src="https://user-images.githubusercontent.com/41648364/114319291-c3aef980-9b08-11eb-8c38-a659c6252a49.png"> | <img src="https://user-images.githubusercontent.com/41648364/114319296-cdd0f800-9b08-11eb-8797-d41a567094dd.png"> |

# Ideas
- Initial icon size could vary based on dpi setting
- Dynamically adjust font resolution when resizing window

# Known issues
- ~Some buttons and icons are slightly mispositioned~ Should all be fixed
- ~Settings app: value picker for Date & Time, Language - tick mark looks wrong~ Fixed
- ~Item/frame/window rounding (notable on lock screen) not scaled correctly~ Should all be fixed

Yunotchi commented on vita3k/gui/src/imgui_impl_sdl_gl3.cpp:
```diff
@@ -406,8 +406,8 @@ IMGUI_API ImTextureID ImGui_ImplSdlGL3_CreateTexture(void *data, int width, int
     glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_R, GL_REPEAT);
-    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
-    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
```
Could this cause problems or undesired effects?

Zangetsu38 replied:
ask @1whatleytay

Zangetsu38 commented:
Squash, no create new commit but edit last and force push

Zangetsu38 commented:
little fail, remove change relate menu_ bar for background, i have set t'his for good reason, here you literally cut and deformed bg lol

Zangetsu38 commented on vita3k/gui/src/content_manager.cpp:
```diff
@@ -255,9 +255,9 @@ void draw_content_manager(GuiState &gui, HostState &host) {
     ImGui::Begin("##content_manager", &gui.live_area.content_manager, ImGuiWindowFlags_NoDecoration | ImGuiWindowFlags_NoSavedSettings);
     if (is_background)
-        ImGui::GetBackgroundDrawList()->AddImage(gui.apps_background["NPXS10026"], ImVec2(0.f, MENUBAR_HEIGHT), display_size);
+        ImGui::GetBackgroundDrawList()->AddImage(gui.apps_background["NPXS10026"], ImVec2(0.f, 0.f), display_size);
```
revert this

Zangetsu38 commented on vita3k/gui/src/live_area.cpp:
```diff
@@ -817,7 +817,7 @@ void draw_live_area_screen(GuiState &gui, HostState &host) {
     if (is_background)
         ImGui::GetBackgroundDrawList()->AddImage((gui.users[host.io.user_id].use_theme_bg && !gui.theme_backgrounds.empty()) ? gui.theme_backgrounds[gui.current_theme_bg] : gui.user_backgrounds[gui.users[host.io.user_id].backgrounds[gui.current_user_bg]],
-            ImVec2(0.f, 32.f), display_size);
+            ImVec2(0.f, 0.f), display_size);
```
keep 32, image obligatoryè start after information bar

Zangetsu38 commented on vita3k/gui/src/settings.cpp:
```diff
@@ -500,9 +500,9 @@ void draw_start_screen(GuiState &gui, HostState &host) {
 
     if (gui.start_background)
-        ImGui::GetForegroundDrawList()->AddImage(gui.start_background, ImVec2(0.f, MENUBAR_HEIGHT), display_size);
+        ImGui::GetBackgroundDrawList()->AddImage(gui.start_background, ImVec2(0.f, 0.f), display_size);
```
rever

Zangetsu38 commented on vita3k/gui/src/settings.cpp:
```diff
@@ -500,9 +500,9 @@ void draw_start_screen(GuiState &gui, HostState &host) {
+        ImGui::GetBackgroundDrawList()->AddImage(gui.start_background, ImVec2(0.f, 0.f), display_size);
     else
-        ImGui::GetForegroundDrawList()->AddRectFilled(ImVec2(0.f, MENUBAR_HEIGHT), display_size, IM_COL32(128.f, 128.f, 128.f, 128.f), 0.f, ImDrawCornerFlags_All);
+        ImGui::GetBackgroundDrawList()->AddRectFilled(ImVec2(0.f, 0.f), display_size, IM_COL32(128.f, 128.f, 128.f, 128.f), 0.f, ImDrawCornerFlags_All);
```
same

Zangetsu38 commented on vita3k/gui/src/trophy_collection.cpp:
```diff
@@ -414,9 +414,9 @@ void draw_trophy_collection(GuiState &gui, HostState &host) {
     ImGui::Begin("##trophy_collection", &gui.live_area.trophy_collection, ImGuiWindowFlags_NoDecoration | ImGuiWindowFlags_NoSavedSettings);
     if (is_background)
-        ImGui::GetBackgroundDrawList()->AddImage(gui.apps_background["NPXS10008"], ImVec2(0.f, MENUBAR_HEIGHT), display_size);
+        ImGui::GetBackgroundDrawList()->AddImage(gui.apps_background["NPXS10008"], ImVec2(0.f, 0.f), display_size);
```
rever

Zangetsu38 commented on vita3k/gui/src/user_management.cpp:
```diff
@@ -237,12 +237,12 @@ void draw_user_management(GuiState &gui, HostState &host) {
         gui.live_area.information_bar = true;
 
-    ImGui::GetBackgroundDrawList()->AddRectFilled(ImVec2(0.f, MENUBAR_HEIGHT), display_size, IM_COL32(10.f, 50.f, 140.f, 255.f), 0.f, ImDrawCornerFlags_All);
+    ImGui::GetBackgroundDrawList()->AddRectFilled(ImVec2(0.f, 0.f), display_size, IM_COL32(10.f, 50.f, 140.f, 255.f), 0.f, ImDrawCornerFlags_All);
```
rever

Zangetsu38 commented on vita3k/gui/src/app_selector.cpp:
```diff
@@ -240,21 +240,21 @@ void draw_app_selector(GuiState &gui, HostState &host) {
     if (is_background)
         ImGui::GetBackgroundDrawList()->AddImage((gui.users[host.io.user_id].use_theme_bg && !gui.theme_backgrounds.empty()) ? gui.theme_backgrounds[gui.current_theme_bg] : gui.user_backgrounds[gui.users[host.io.user_id].backgrounds[gui.current_user_bg]],
-            ImVec2(0.f, MENUBAR_BG_HEIGHT), display_size);
+            ImVec2(0.f, 0.f), display_size);
```
revert

Zangetsu38 commented on vita3k/gui/src/app_selector.cpp:
```diff
@@ -240,21 +240,21 @@ void draw_app_selector(GuiState &gui, HostState &host) {
+            ImVec2(0.f, 0.f), display_size);
     else
-        ImGui::GetBackgroundDrawList()->AddRectFilled(ImVec2(0.f, MENUBAR_BG_HEIGHT), display_size, IM_COL32(11.f, 90.f, 252.f, 160.f), 0.f, ImDrawCornerFlags_All);
+        ImGui::GetBackgroundDrawList()->AddRectFilled(ImVec2(0.f, 0.f), display_size, IM_COL32(11.f, 90.f, 252.f, 160.f), 0.f, ImDrawCornerFlags_All);
```
same

Zangetsu38 commented on vita3k/gui/src/controls_dialog.cpp:
```diff
@@ -61,11 +61,13 @@ static void remapper_button(GuiState &gui, HostState &host, int *button, const c
     ImGui::SetNextWindowPos(ImVec2(ImGui::GetIO().DisplaySize.x / 2.f, ImGui::GetIO().DisplaySize.y / 2.f), ImGuiCond_Always, ImVec2(0.5f, 0.5f));
     ImGui::Begin("Controls", &gui.controls_menu.controls_dialog);
+    //ImGui::SetNextWindowPos(ImVec2(ImGui::GetIO().DisplaySize.x / 2.f, ImGui::GetIO().DisplaySize.y / 2.f), ImGuiCond_Always, ImVec2(0.5f, 0.5f));
+    //ImGui::Begin("Controls", &gui.controls_menu.controls_dialog, ImGuiWindowFlags_AlwaysAutoResize);
```
🤔  ?

Yunotchi replied:
I will fix this lol

Yunotchi commented:
> psvita theme bg is 960x512, because is 544- Information bar (32)

The background images in `AppData\Roaming\Vita3K\Vita3K\vs0\app` (e.g. `\NPXS10002\sce_sys\pic0.png`) are all 960x544, and official wallpapers provided by game companies are 960x544 (e.g. https://aquaplus.jp/dt2vita/wallpaper/) 

Are the themes that you buy from PSN different?

Zangetsu38 commented:
> 
> 
> > psvita theme bg is 960x512, because is 544- Information bar (32)
> 
> The background images in `AppData\Roaming\Vita3K\Vita3K\vs0\app` (e.g. `\NPXS10002\sce_sys\pic0.png`) are all 960x544, and official wallpapers provided by game companies are 960x544 (e.g. https://aquaplus.jp/dt2vita/wallpaper/)
> 
> Are the themes that you buy from PSN different?

ah yes sorry, i forget i u_sing app bg for system app lol, is like splash screen, so yes is 544 but, for theme is 512, so for start screen and app selector keep menubar

Zangetsu38 commented:
also on psvita, if you using personal bg for start screen on wallpaper, it obligatory resize image for start after information bar, for no cut image and no deforme, is for that on psvita, it crop image

Yunotchi commented:
Ah, thank you for explaining
I'll take another look at this soon :)

Zangetsu38 commented:
> 
> 
> Ah, thank you for explaining
> I'll take another look at this soon :)

i have remove my comment on change so no request,  so just loop where is keep ^^

Zangetsu38 commented on vita3k/app/src/app_init.cpp:
```diff
@@ -151,7 +151,10 @@ bool init(HostState &state, Config &cfg, const Root &root_paths, CPUDepInject in
-    state.window = WindowPtr(SDL_CreateWindow(window_title, SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, DEFAULT_RES_WIDTH, DEFAULT_RES_HEIGHT, window_type | SDL_WINDOW_RESIZABLE | SDL_WINDOW_ALLOW_HIGHDPI), SDL_DestroyWindow);
+    float ddpi, hdpi, vdpi;
+    SDL_GetDisplayDPI(0, &ddpi, &hdpi, &vdpi);
+    float dpi_scale = ddpi / 96;
```
try keep this value here for no repeat this code in other part, with adding GuiState or with store value inside state (host)

Zangetsu38 commented on vita3k/WindowsDpiAwareness.manifest:
```diff
@@ -0,0 +1,9 @@
+      <dpiAwareness xmlns="http://schemas.microsoft.com/SMI/2016/WindowsSettings">PerMonitorV2</dpiAwareness>
+    </asmv3:windowsSettings>
+  </asmv3:application>
+</assembly>
```
never forget add new line in end file

Zangetsu38 commented on vita3k/gui/src/app_selector.cpp:
```diff
@@ -184,13 +184,13 @@ static float scroll_type, current_scroll_pos, max_scroll_pos;
     const auto scal = ImVec2(display_size.x / 960.0f, display_size.y / 544.0f);
-    const auto MENUBAR_HEIGHT = 32.f * scal.y;
+    const auto INFORMATION_BAR_HEIGHT = 32.f * scal.y;
     const auto MENUBAR_BG_HEIGHT = !gui.live_area.information_bar ? 22.f : 32.f * scal.y;
```
AH my bad lol if yçou can change this to 
```suggestion
    const auto MENUBAR_BG_HEIGHT = !gui.live_area.information_bar ? 22.f : INFORMATION_BAR_HEIGHT ;
```
thx

Yunotchi commented:
Made a few changes:
- DPI awareness is done in cmake with manifest file -- this is better than doing it at runtime, and now the application loads a higher resolution version of its icon
- Reverted changes to backgrounds -- I want to look at this again later
- Notification icon replaced with more authentic circle
- Center user names in User Management
- DPI scaling is done with `ImGuiIO.DisplayFramebufferScale`. Fonts are loaded at a larger size and scaled down globally via `ImGuiIO.FontGlobalScale`
- Renamed `MENUBAR_HEIGHT` to `INFORMATION_BAR_HEIGHT` in the gui classes

Zangetsu38 commented on vita3k/gui/src/information_bar.cpp:
```diff
@@ -356,7 +356,7 @@ static void draw_notice_info(GuiState &gui, HostState &host) {
             ImGui::GetForegroundDrawList()->AddImage(gui.theme_information_bar_notice["no"], NOTICE_POS, ImVec2(NOTICE_POS.x + NOTICE_SIZE.x, NOTICE_SIZE.y));
         else
-            ImGui::GetForegroundDrawList()->AddRectFilled(ImVec2(display_size.x - (70.f * SCAL.x), (-24.f * SCAL.y)), ImVec2(display_size.x + (14.f * SCAL.y), 60.f * SCAL.y), IM_COL32(62.f, 98.f, 160.f, 255.f), 75.f * SCAL.x, ImDrawCornerFlags_All);
+            ImGui::GetForegroundDrawList()->AddCircleFilled(ImVec2(display_size.x - (28.f * SCAL.x), (18.f * SCAL.x)), 44.0f * SCAL.x, IM_COL32(62.f, 98.f, 160.f, 255.f));
```
i have try is great :D, add same on 353 line just before, is notif new, size is little more big try calculate for get same position :)

Yunotchi commented:
Made some more changes, hopefully final for this PR:
- Added scaling for save dialog windows
- Fixed button rounding in many places not scaling correctly with window size
- Reverted changes to notification icon -- I will do this and other GUI tweaks in another PR
- Live area: tweaked start button gate's rounding and thickness to look more authentic
- App selector: icons now scale with window size
- Content manager: black bar at bottom now fills the whole window's width

pent0 commented on vita3k/gui/src/gui.cpp:
```diff
@@ -105,7 +107,16 @@ static void init_style() {
+
+#ifdef _WIN32
+    if (host.dpi_scale > 1) {
+        gui.monospaced_font = io.Fonts->AddFontFromFileTTF("C:\\Windows\\Fonts\\consola.ttf", 13.0f * host.dpi_scale * host.dpi_scale, NULL, io.Fonts->GetGlyphRangesJapanese()); // multiply by dpi_scale twice to correctly scale
```
Constant string should be made a separate variable with name explaining the usage

pent0 commented on vita3k/gui/src/information_bar.cpp-d13cf8a6:
```diff
@@ -0,0 +1,562 @@
+﻿// Vita3K emulator project
```
This file is unintentionally left over?

pent0 reviewed (changes requested):
I can't judge much about how you implement it, there are still some more nits

Yunotchi commented:
Thank you for the feedback, pent0. I've addressed the issues you raised.

I have also reverted the implementation of DPI-awareness I made in #1220 because:
- shcore isn't available in Windows 7
- as I mentioned in an above comment, implementing DPI-awareness via an application manifest is better

scribam commented:
I am not sure if it can be useful but there is a function in ImGui about DPI awareness : https://github.com/ocornut/imgui/blob/7b8932554e04aca62d3c2d4ffbb5b9212a55f801/backends/imgui_impl_win32.cpp#L444

Yunotchi commented:
Hm, I'll look at that ImGui function to see what it does.

My current implementation is working correctly on Windows 10 -- see the last column in this screenshot:
![image](https://user-images.githubusercontent.com/41648364/116457298-21f31080-a85b-11eb-9627-a83f2b836c45.png)

I don't have a Windows 7 machine to test with but [Microsoft's documentation](https://docs.microsoft.com/en-us/windows/win32/hidpi/setting-the-default-dpi-awareness-for-a-process) says it should work on all Windows since Vista.

sunho commented:
Can you rebase this pr.

Zangetsu38 commented on vita3k/gui/src/common_dialog.cpp:
```diff
@@ -60,18 +61,22 @@ static void draw_ime_dialog(DialogState &common_dialog) {
-    const ImVec2 BUTTON_SIZE = ImVec2(200.f, 35.f);
-    const ImVec2 WINDOW_SIZE = ImVec2(ImGui::GetIO().DisplaySize.x / 1.7f, ImGui::GetIO().DisplaySize.y / 1.5f);
-    const ImVec2 PROGRESS_BAR_SIZE = ImVec2(WINDOW_SIZE.x - 120.f, 7.f);
+    LOG_INFO("draw_message_dialog");
```
Hmmmmm ?

Yunotchi replied:
Darn it, will fix

Yunotchi replied:
Removed

pent0 reviewed (dismissed):
About the Linear change, it was for font blurry I think, but if you do DPI scale I don't think it should be a worry to not use Nearest anymore. But take notes if anything degrades (doubts).

Edit: Rebase then approve again.

Yunotchi commented:
OK, I've rebased my branch.

> About the Linear change, it was for font blurry I think, but if you do DPI scale I don't think it should be a worry to not use Nearest anymore. But take notes if anything degrades (doubts).

My reason for changing from nearest neighbour to linear was for icons:
<img width="230" alt="1 5DPI-windowed-2x" src="https://user-images.githubusercontent.com/41648364/116557894-cd04d800-a8f6-11eb-8494-2e62a295508d.png">
